File: everything_parsing/09-input-buffering/java/Scanner.java
import java.util.ArrayList;
import java.util.List;

/**
 * A deliberately tiny scanner, built ON TOP of Buffer, to exercise the buffer's
 * peek / advance / mark / retract. Recognises:
 *
 *   WORD   [a-z]+
 *   NUM    [0-9]+
 *   FLOAT  [0-9]+ '.' [0-9]+
 *   PLUS   '+'
 *   (spaces are skipped)
 *
 * The FLOAT rule is where retract earns its keep: after reading "12." the
 * scanner must peek one more character; if it isn't a digit, the '.' belongs to
 * the next token and `forward` retracts one.
 */
public final class Scanner {
    public record Tok(String kind, String text) {}

    static boolean isDigit(int c) { return c >= '0' && c <= '9'; }
    static boolean isLower(int c) { return c >= 'a' && c <= 'z'; }

    /** The next token, or null at end of input. `retracts` is incremented on each retract. */
    static Tok next(Buffer b, int[] retracts) {
        while (b.peek() == ' ') b.advance();
        b.mark();
        int c = b.peek();
        if (c == Buffer.EOF) return null;

        if (isLower(c)) {
            do { b.advance(); } while (isLower(b.peek()));
            return new Tok("WORD", b.lexeme());
        }
        if (isDigit(c)) {
            do { b.advance(); } while (isDigit(b.peek()));
            if (b.peek() == '.') {
                b.advance();                       // tentatively take the '.'
                if (isDigit(b.peek())) {
                    do { b.advance(); } while (isDigit(b.peek()));
                    return new Tok("FLOAT", b.lexeme());
                }
                b.retract(1);                      // '.' was not part of the number
                retracts[0]++;
                return new Tok("NUM", b.lexeme());
            }
            return new Tok("NUM", b.lexeme());
        }
        if (c == '+') { b.advance(); return new Tok("PLUS", b.lexeme()); }

        b.advance();
        return new Tok("ERR", b.lexeme());
    }

    static List<Tok> scanAll(Buffer b, int[] retracts) {
        List<Tok> out = new ArrayList<>();
        Tok t;
        while ((t = next(b, retracts)) != null) out.add(t);
        return out;
    }
}

File: everything_parsing/09-input-buffering/js/scanner.mjs
// A tiny scanner built on Buffer, to exercise peek / advance / mark / retract.
//   WORD [a-z]+   NUM [0-9]+   FLOAT [0-9]+ '.' [0-9]+   PLUS '+'   (spaces skipped)
// The FLOAT rule is where retract is needed: after "12." peek one more char;
// if it isn't a digit, the '.' belongs to the next token and forward retracts one.

import { EOF } from "./buffer.mjs";

const isDigit = (c) => c >= 48 && c <= 57;
const isLower = (c) => c >= 97 && c <= 122;

export function next(b, retracts) {
  while (b.peek() === 32) b.advance();
  b.mark();
  const c = b.peek();
  if (c === EOF) return null;

  if (isLower(c)) {
    do { b.advance(); } while (isLower(b.peek()));
    return { kind: "WORD", text: b.lexeme() };
  }
  if (isDigit(c)) {
    do { b.advance(); } while (isDigit(b.peek()));
    if (b.peek() === 46) { // '.'
      b.advance();
      if (isDigit(b.peek())) {
        do { b.advance(); } while (isDigit(b.peek()));
        return { kind: "FLOAT", text: b.lexeme() };
      }
      b.retract(1);
      retracts[0]++;
      return { kind: "NUM", text: b.lexeme() };
    }
    return { kind: "NUM", text: b.lexeme() };
  }
  if (c === 43) { b.advance(); return { kind: "PLUS", text: b.lexeme() }; } // '+'

  b.advance();
  return { kind: "ERR", text: b.lexeme() };
}

export function scanAll(b, retracts) {
  const out = [];
  let t;
  while ((t = next(b, retracts)) !== null) out.push(t);
  return out;
}
